File: project_euler/001-multiples_of_3_and_5.rb
# If we list all the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6 and 9. The sum of these multiples is 23.
# Find the sum of all the multiples of 3 or 5 below 1000.

class Multiples
  attr_accessor :max
  attr_reader :multiples

  def initialize( max )
    @max = max
    @multiples = []
  end

  def sum_of_multiples_of_three_and_five
    multiples_of_three_and_five
    multiples.inject(:+)
  end

  def multiples_of_three_and_five
    (1...max).each do |num|
      multiples.push(num) && next if is_a_multiple_of?(num, 3)
      multiples.push(num) if is_a_multiple_of? num, 5
    end
  end

  def is_a_multiple_of?( num, value )
    num % value == 0
  end
end

mul = Multiples.new( 1000 )
puts mul.sum_of_multiples_of_three_and_five

File: project_euler/001-multiples_of_3_and_5.js
// If we list all the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6 and 9. The sum of these multiples is 23.
// Find the sum of all the multiples of 3 or 5 below 1000.

var Multiples = function(max) {
  var multiples = [];
  var num;

  function multiplesOfThreeOrFive(num) {
    if (num % 3 == 0) {
      multiples.push(num)
      return;
    } else if (num % 5 == 0) {
      multiples.push(num)
      return;
    } else {
      return;
    }
  }

  function calculateAllMultiples() {
    for (i = 1; i < max; i++) {
      multiplesOfThreeOrFive(i);
    }
    return multiples;
  }

  function addMultiples() {
    total = 0;
    for (i = 0; i < multiples.length; i++) {
      total += multiples[i];
    }
    return total;
  }

  calculateAllMultiples();
  return addMultiples();
}

console.log(Multiples(1000));
